packages/lib/gpc/src/gpcCompile.ts

```diff
@@ -360,42 +362,64 @@ function compileProofListMembership<
   paramMaxListElements: number
 ): {
   listComparisonValueIndex: CircuitSignal[];
+  listContainsComparisonValue: CircuitSignal[];
   listValidValues: CircuitSignal[][];
 } {
   // Arrange list names alphabetically.
   const listNameOrder = Object.keys(listConfig).sort();
 
-  // Do the same for the lists of comparison IDs and string them together.
+  // Arrange for membership before non-membership checks and consider the
+  // lists of comparison IDs in alphabetical order, stringing them all together.
   // This takes into account the possibility where different entry (or tuple)
   // values must lie in the same list, or the case one entry (or tuple) value
   // must lie in multiple lists.
-  const listIdPairs = listNameOrder
-    .map((listName: PODName): [PODName, PODEntryIdentifier][] =>
-      listConfig[listName].sort().map((elementId) => [listName, elementId])
+  const listIndicatorIdTriples = listNameOrder.flatMap((listName) =>
+    (Object.keys(listConfig[listName]).sort() as ListMembershipEnum[]).flatMap(
+      (membershipIndicator) =>
+        listConfig[listName][membershipIndicator]
+          .sort()
+          .flatMap((elementId) => [
+            [listName, membershipIndicator, elementId] as [
+              PODName,
+              ListMembershipEnum,
+              PODEntryIdentifier | TupleIdentifier
+            ]
+          ])
     )
-    .flat();
+  );
 
   // Compile listComparisonValueIndex
-  const unpaddedListComparisonValueIndex = listIdPairs.map((listIdPair) => {
-    const [listName, elementId] = listIdPair;
-
-    const idx = isTupleIdentifier(elementId)
-      ? tupleMap.get(elementId as TupleIdentifier)?.tupleIndex
-      : entryMap.get(elementId)?.entryIndex;
+  const unpaddedListComparisonValueIndex = listIndicatorIdTriples.map(
+    (listIndIdTriple) => {
+      const [listName, _indicator, elementId]: [
```

Comment:
Can you do this destructuring in the argument declaration above?  That would be clearer and terser.
If this were a for loop instead of a map I know you could to the destructuring inline.